packages/modular-scripts/src/build.ts

```diff
@@ -0,0 +1,618 @@
+      },
+      error: (...args: Parameters<typeof console.error>) => {
+        return console.error('$ ' + directoryName + ':', ...args);
+      },
+    };
+  }
+  return consoles[directoryName];
+}
+
+// from https://github.com/Microsoft/TypeScript/issues/6387
+// a helper to output a readable message from a ts diagnostics object
+function reportTSDiagnostics(
+  directoryName: string,
+  diagnostics: ts.Diagnostic[],
+): void {
+  diagnostics.forEach((diagnostic) => {
+    let message = 'Error';
+    if (diagnostic.file) {
+      const where = diagnostic.file.getLineAndCharacterOfPosition(
+        diagnostic.start as number,
+      );
+      message += ` ${diagnostic.file.fileName} ${where.line}, ${
+        where.character + 1
+      }`;
+    }
+    message +=
+      ': ' + ts.flattenDiagnosticMessageText(diagnostic.messageText, '\n');
+    console.log(message);
+  });
+}
+
+const extensions = ['.ts', '.tsx', '.js', '.jsx'];
+const outputDirectory = 'dist';
+const typescriptConfigFilename = 'tsconfig.json';
+const packagesRoot = 'packages';
+
+// list of all directories under packages
+const packageDirectoryNames = fse
+  .readdirSync(packagesRoot, { withFileTypes: true })
+  .filter((directoryEntry) => directoryEntry.isDirectory())
+  .map((directory) => directory.name);
+
+// dependencies defined at the root
+const rootPackageJsonDependencies =
+  (fse.readJSONSync('package.json') as PackageJson).dependencies || {};
+
+// a map of all package.json contents, indexed by package name
+const packageJsons: { [key: string]: PackageJson } = {};
+// a map of all package.json contents, indexed by directory name
+const packageJsonsByDirectoryName: {
+  [key: string]: PackageJson;
+} = {};
+// an array of all the package names
+const packageNames: string[] = [];
+
+// let's populate the above three
+for (let i = 0; i < packageDirectoryNames.length; i++) {
+  const pathToPackageJson = path.join(
+    packagesRoot,
+    packageDirectoryNames[i],
+    'package.json',
+  );
+  if (fse.existsSync(pathToPackageJson)) {
+    const packageJson = fse.readJsonSync(pathToPackageJson) as PackageJson;
+    if (!packageJson.name) {
+      continue;
+    }
+    packageJsons[packageJson.name] = packageJson;
+    packageJsonsByDirectoryName[packageDirectoryNames[i]] = packageJson;
+    packageNames.push(packageJson.name);
+  }
+}
+
+// TODO: do a quick check to make sure workspaces aren't
+// explicitly included in dependencies
+// maybe that belongs in `modular check`
+
+const publicPackageJsons: {
+  [name: string]: PackageJson;
+} = {};
+
+function distinct<T>(arr: T[]): T[] {
+  return [...new Set(arr)];
+}
+
+const typescriptConfig: TSConfig = {};
+// validate tsconfig
+{
+  // Extract configuration from config file and parse JSON,
+  // after removing comments. Just a fancier JSON.parse
+  const result = ts.parseConfigFileTextToJson(
+    typescriptConfigFilename,
+    fse.readFileSync(typescriptConfigFilename, 'utf8').toString(),
```

Comment:
Perhaps these reads should be prefixed with `getModularRoot`?

Reply:
also good catch, will do

Reply:
Ok, instead, I've made this script error if it's not run from modular root. Makes it more broadly applicable across this file.

Reply:
LGTM